internal/resolver/dns/dns_resolver_test.go

```diff
@@ -764,103 +764,113 @@ func TestDNSResolverExponentialBackoff(t *testing.T) {
-		// Cause ClientConn to return an error.
-		cc.updateStateErr = balancer.ErrBadResolverState
-		r, err := b.Build(resolver.Target{Endpoint: a.target}, cc, resolver.BuildOptions{})
-		if err != nil {
-			t.Fatalf("%v\n", err)
-		}
-		var state resolver.State
-		var cnt int
-		for i := 0; i < 2000; i++ {
-			state, cnt = cc.getState()
-			if cnt > 0 {
-				break
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			b := NewBuilder()
+			cc := &testClientConn{target: test.target}
+			// Cause ClientConn to return an error.
+			cc.updateStateErr = balancer.ErrBadResolverState
+			r, err := b.Build(resolver.Target{Endpoint: test.target}, cc, resolver.BuildOptions{})
+			if err != nil {
+				t.Fatalf("%v\n", err)
 			}
-			time.Sleep(time.Millisecond)
-		}
-		if cnt == 0 {
-			t.Fatalf("UpdateState not called after 2s; aborting")
-		}
-		if !reflect.DeepEqual(a.addrWant, state.Addresses) {
-			t.Errorf("Resolved addresses of target: %q = %+v, want %+v", a.target, state.Addresses, a.addrWant)
-		}
-		sc := scFromState(state)
-		if a.scWant != sc {
-			t.Errorf("Resolved service config of target: %q = %+v, want %+v", a.target, sc, a.scWant)
-		}
-		// Cause timer to go off 10 times, and see if it calls updateState() correctly.
-		for i := 0; i < 10; i++ {
-			timer := <-timerChan
-			timer.Reset(0)
-		}
-		// Poll to see if DNS Resolver updated state the correct number of times, which allows time for the DNS Resolver to call
-		// ClientConn update state.
-		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-		for {
-			timer := time.NewTimer(time.Millisecond)
-			select {
-			case <-ctx.Done():
-				t.Fatalf("Exponential backoff is not working as expected - should update state 11 times instead of %d", cc.updateStateCalls)
-			case <-timer.C:
+			var state resolver.State
+			var cnt int
+			for i := 0; i < 2000; i++ {
+				state, cnt = cc.getState()
+				if cnt > 0 {
+					break
+				}
+				time.Sleep(time.Millisecond)
 			}
-			cc.m1.Lock()
-			if cc.updateStateCalls == 11 {
-				cc.m1.Unlock()
-				break
+			if cnt == 0 {
+				t.Fatalf("UpdateState not called after 2s; aborting")
 			}
-			cc.m1.Unlock()
-		}
-		cancel()
+			if !reflect.DeepEqual(test.addrWant, state.Addresses) {
+				t.Errorf("Resolved addresses of target: %q = %+v, want %+v", test.target, state.Addresses, test.addrWant)
+			}
+			sc := scFromState(state)
+			if test.scWant != sc {
+				t.Errorf("Resolved service config of target: %q = %+v, want %+v", test.target, sc, test.scWant)
+			}
+			// Cause timer to go off 10 times, and see if it calls updateState() correctly.
+			for i := 0; i < 10; i++ {
+				timer := <-timerChan
+				timer.Reset(0)
+			}
+			// Poll to see if DNS Resolver updated state the correct number of times, which allows time for the DNS Resolver to call
+			// ClientConn update state.
+			deadline := time.Now().Add(defaultTestTimeout)
+			for {
+				cc.m1.Lock()
+				got := cc.updateStateCalls
+				if got == 11 {
```

Comment:
`Unlock` before the `if` to save a line of code and help prevent refactorings from forgetting to unlock.

Reply:
Changed.